metric_test.go

```diff
@@ -0,0 +1,223 @@
+		},
+		{
+			input: "testing|ST[blah:blah]",
+			tok:   tokenMetric,
+			lit:   "testing",
+		},
+		{
+			input: "t|esting|ST[blah:blah]",
+			tok:   tokenMetric,
+			lit:   "t|esting",
+		},
+		{
+			input: "t|Sesting|ST[blah:blah]",
+			tok:   tokenMetric,
+			lit:   "t|Sesting",
+		},
+		{
+			input: "tel|a|M|Ssting|ST[blah:blah]",
+			tok:   tokenMetric,
+			lit:   "tel|a|M|Ssting",
+		},
+		{
+			input: "testing|ST[blah:blah]",
+			tok:   tokenMetric,
+			lit:   "testing",
+		},
+	}
+
+	for _, c := range cases {
+		buf := bytes.NewBufferString(c.input)
+		s := newMetricScanner(buf)
+		tok, lit, err := s.scanMetricName()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if tok != c.tok {
+			t.Error("failed to find the metric ident")
+		}
+		if lit != c.lit {
+			t.Error("incorrect literal scanned: ", c.lit, lit)
+		}
+	}
+}
+
+func TestScanTagSep(t *testing.T) {
+	stBuf := bytes.NewBufferString("|ST[blah:blah]")
+	s := newMetricScanner(stBuf)
+	tok, lit, err := s.peekTagSep()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if tok != tokenStreamTag {
+		t.Error("expected stream tag separator")
+	}
+
+	if lit != "|ST" {
+		t.Error("incorrect literal scanned: ", lit)
+	}
+
+	tok, lit, err = s.scanTagSep()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if tok != tokenStreamTag {
+		t.Error("expected stream tag separator")
+	}
+
+	if lit != "|ST" {
+		t.Error("incorrect literal scanned: ", lit)
+	}
+
+	mtBuf := bytes.NewBufferString("|MT[blah:blah]")
+	s = newMetricScanner(mtBuf)
+	tok, lit, err = s.peekTagSep()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if tok != tokenMeasurementTag {
+		t.Error("expected measurement tag separator")
+	}
+
+	if lit != "|MT" {
+		t.Error("incorrect literal scanned: ", lit)
+	}
+
+	tok, lit, err = s.scanTagSep()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if tok != tokenMeasurementTag {
+		t.Error("expected measurement tag separator")
+	}
+
+	if lit != "|MT" {
+		t.Error("incorrect literal scanned: ", lit)
+	}
+}
+
+func TestMetricParser(t *testing.T) {
+	var cases = []struct {
+		input string // input
+		num   int    // number of derived stream tags
+		lit   string // metric name literal
+	}{
+		{
+			input: "testing",
+			num:   0,
+			lit:   "testing",
+		},
+		{
+			input: "testing*",
+			num:   0,
+			lit:   "testing*",
+		},
+		{
+			input: "testing|ST[blah:blah]",
+			num:   1,
+			lit:   "testing",
+		},
+		{
+			input: "t|esting|ST[blah:blah]",
+			num:   1,
+			lit:   "t|esting",
+		},
+		{
+			input: "t|Sesting|ST[blah:blah]",
+			num:   1,
+			lit:   "t|Sesting",
+		},
+		{
+			input: "tel|a|M|Ssting|ST[blah:blah]",
+			num:   1,
+			lit:   "tel|a|M|Ssting",
+		},
+		{
+			input: "testing|ST[blah:blah]",
+			num:   1,
+			lit:   "testing",
+		},
+		{
+			input: "testing|ST[blah:blah,blah1:blah,blah2:blah]",
+			num:   3,
+			lit:   "testing",
+		},
+		{
+			input: `testing|ST[b"QUFB":blah,blah1:b"QkJCQgo=",b"YWFh":b"YmJi"]`,
+			num:   3,
+			lit:   "testing",
+		},
+		{
+			input: "testing|ST[blah:blah:blah:blah]",
```

Comment:
```suggestion
			input: "testing|ST[blah:blah:blah:blah]",
```
This test isn't actually valid as this is not a valid tag format.

Reply:
This has been corrected.